### PR TITLE
bring latest changes to master

### PR DESCRIPTION
**Pull request by bogdan-rosianu**

### CODE REVIEW

File: src/signTxHash.c [hunk 1]
```diff
@@ -238,6 +251,9 @@ uint16_t process_field(void) {
     err = verify_version(&valid_field);
     if (err != MSG_OK)
         return err;
+    err = verify_options(&valid_field);
+    if (err != MSG_OK)
+        return err;    
 
     // verify the rest of the fields that are not displayed
     valid_field |= strncmp(tx_hash_context.current_field, NONCE_FIELD, strlen(NONCE_FIELD)) == 0;
```
Comment (mihaib79) [inline src/signTxHash.c:259]: We should add here checks for username fields (sender / receiver).

File: src/viewAppVersion.c [hunk 2]
```diff
@@ -2,6 +2,7 @@
 #include "os.h"
 #include "ux.h"
 #include "utils.h"
+#include "menu.h"
```
Comment (AdoAdoAdo) [inline src/viewAppVersion.c:5]: why is this required? there is no other code change in this file

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: src/main.c [hunk 3]
```diff
@@ -91,7 +90,8 @@ void handleApdu(volatile unsigned int *flags, volatile unsigned int *tx) {
                     break;
 
                 case INS_SIGN_TX:
-                    handleSignTx(G_io_apdu_buffer[OFFSET_P1], G_io_apdu_buffer[OFFSET_P2], G_io_apdu_buffer + OFFSET_CDATA, G_io_apdu_buffer[OFFSET_LC], flags, tx);
+                    // sign tx is deprecated is this version and hash signing should be used
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply src/main.c:93]: done

File: src/signMsg.c [hunk 4]
```diff
@@ -1,4 +1,5 @@
 #include "signMsg.h"
+#include "utils.h"
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply src/signMsg.c:2]: it was a lint issue fixed by the Ledger team so we merged those changes into our repo

File: src/signTxHash.c [hunk 5]
```diff
@@ -228,7 +244,7 @@ uint16_t process_field(void) {
         return err;
     err = verify_gaslimit(&valid_field);
     if (err != MSG_OK)
-        return err;
+        return err;
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply src/signTxHash.c:247]: done

File: src/signTxHash.c [hunk 6]
```diff
@@ -299,19 +321,33 @@ uint16_t parse_data(uint8_t *dataBuffer, uint16_t dataLength) {
                 break;
             case JSON_PROCESSING_STRING_VALUE:
                 if (c == '"') {
+                    tx_hash_context.data_field_size = tx_hash_context.current_value_len / 4 * 4;
+                    tx_hash_context.data_field_size = tx_hash_context.data_field_size / 4 * 3;
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply src/signTxHash.c:325]: for converting the `base64` encoded data length into decoded `string` length. Same logic as before, only moved

File: src/utils.c [hunk 7]
```diff
@@ -199,18 +199,34 @@ bool is_digit(char c) {
...
-    uint128_t fee128;
+bool gas_to_fee(uint64_t gas_limit, uint64_t gas_price, uint32_t data_size, char *fee, size_t size)
+{
+    uint128_t x = {0, GAS_PER_DATA_BYTE};
+    uint128_t y = {0, data_size};
+    uint128_t z;
+    uint128_t gas_unit_for_move_balance; 
+    // tx fee formula
+    // gasUnitForMoveBalance := (minGasLimit + len(data)*gasPerDataByte)
+    // txFEE = gasUnitForMoveBalance * GASPRICE + (gasLimit - gasUnitForMoveBalance) * gasPriceModifier * GASPRICE
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply src/utils.c:210]: updated comment

File: src/viewAppVersion.c (hunk 2 above)
Comment (bogdan-rosianu) [reply src/viewAppVersion.c:5]: same reason as above. linter issue found by Ledger team

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

**Review by mihaib79 — DISMISSED** [review]

File: src/signTxHash.c [hunk 8]
```diff
@@ -319,27 +318,30 @@ uint16_t parse_data(uint8_t *dataBuffer, uint16_t dataLength) {
...
-                    tx_hash_context.data_field_size = tx_hash_context.current_value_len / 4 * 4;
-                    tx_hash_context.data_field_size = tx_hash_context.data_field_size / 4 * 3;
-                    if(penultimate_data_char == '=') {
-                        tx_hash_context.data_field_size--;
-                    }
-                    if(last_data_char == '=') {
-                        tx_hash_context.data_field_size--;
+                    if (isDataField) {
+                        uint32_t data_value_len;
+                        data_value_len = tx_hash_context.current_value_len / 4 * 4;
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply src/signTxHash.c]: done

File: src/signTxHash.c [hunk 9]
```diff
@@ -319,27 +318,30 @@ uint16_t parse_data(uint8_t *dataBuffer, uint16_t dataLength) {
...
+                        data_value_len = data_value_len / 4 * 3;
+                        //  remove trailing padding chars from count if any
+                        if (tx_hash_context.current_value_len > 2) {
+                            if(dataBuffer[idx-2] == '='){
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply src/signTxHash.c:334]: added a comment to explain the reasoning

**Review by AdoAdoAdo — APPROVED** [review]

**Review by mihaib79 — APPROVED** [review]

**Pull request merged**